### ISSUE TITLE
MQTT connection through proxy

### ISSUE THREAD

**Issue by oikibun** [thin-edge/thin-edge.io#3228]
Traditional company or factories is limited the Internet connection using proxy.

HTTPS can set that setting like below in an env file.
"HTTPS_PROXY=http://proxy.foobar.co.jp"
But the MQTTS doesn't have such an option and can't connect to the Internet side(Cumulocity IoT, AWS).

It's helpful for the old company to over-proxy connection.

**Comment by reubenmiller** [comment]
We'll have to investigate the feasibility of this, though one thing that seems clear is that supporting a connection through a http proxy would only be possible if MQTT over websockets is used (not MQTT TCP, which is currently the only supported MQTT transport mechanism)

**Comment by TyrManuZ** [comment]
@reubenmiller 
There is a standard called HTTP CONNECT (https://en.wikipedia.org/wiki/HTTP_tunnel#HTTP_CONNECT_tunneling) and most MQTT client libraries also support it (e.g. paho). This will send an initial HTTP request to the proxy, telling it to open a TCP connection to the target and start tunneling. This standard is agnostic to the underlying protocol and you could run normal MQTT through it.
The limitations that I see is that
1. The proxy server needs to support it as well
2. The proxy server configurations can still harm your experience with short lifetimes of persistent connections etc.

Mosquitto bridge does not support such a mechanism

**Comment by reubenmiller** [comment]
> @reubenmiller There is a standard called HTTP CONNECT (https://en.wikipedia.org/wiki/HTTP_tunnel#HTTP_CONNECT_tunneling) and most MQTT client libraries also support it (e.g. paho). This will send an initial HTTP request to the proxy, telling it to open a TCP connection to the target and start tunneling. This standard is agnostic to the underlying protocol and you could run normal MQTT through it. The limitations that I see is that
> 
> 1. The proxy server needs to support it as well
> 2. The proxy server configurations can still harm your experience with short lifetimes of persistent connections etc.
> 
> Mosquitto bridge does not support such a mechanism

yes that **may be** doable, but before we can even start experimenting with it, we need some kind of setup verify that it would work, and that setup would have to faithfully represent the desired proxy in the end (but no one seems to be able to tell us of some proxy services which are desired to be used).

In past unrelated projects, I have had some success with using [proxychains-ng](https://github.com/rofl0r/proxychains-ng), though it was always a bit of a hassle to get it correctly configured and debugging was a nightmare, but it could be a possible way forward if you wanted to experiment with some local setup, but it will only work with glibc compiled applications (which by default, thin-edge.io is not, though you can compile thin-edge.io yourself to use glibc instead of musl)

**Comment by reubenmiller** [comment]
Or some other possibilities (after doing a bit more research):
* socat to setup a bi-directional data channel to forward to a HTTP proxy (only for experts)
* https://www.stunnel.org/
* https://github.com/go-gost/gost

**Comment by reubenmiller** [comment]
@TyrManuZ Regarding usine `HTTP CONNECT` for tunnelling. Do we know if it is common practice that proxy services offer this feature? As I've seen the default [squid proxy](https://www.squid-cache.org/) settings don't actually allow this.

So it would be really helpful if anyone could offer some examples about proxy services being used and whether they support `HTTP CONNECT` tunnelling.

**Comment by jarhodes314** [comment]
I've done some very quick investigations into this, and provided the proxy supports HTTP CONNECT, this should be very simple to support via the built-in bridge. rumqttc (the mqtt library we use internally) supports HTTP CONNECT proxying via the [async-http-proxy](https://docs.rs/async-http-proxy) crate, and I've verified manually that I can connect to a locally running MQTT broker via a `go-gost` proxy simply by enabling the proxy feature and configuring the proxy details with rumqttc and connecting as normal from there.

**Comment by TyrManuZ** [comment]
I checked the setup where the request orginially came from that Fortinet proxy is used. From some local tcpdumps I can see other services running on the machine where thin-edge is installed that actually communicate through the proxy by using HTTP CONNECT.

**Comment by reubenmiller** [comment]
The change will be in the next official release, 1.5.0 which is due at the start of next week (6th May 2025).

But since the change has been merged, below should be a good starting guide until we have official docs available online.

**Pre-requisites**

Assuming you have already configured the `c8y.url` and the device's certificate is trusted by the platform. If so, then you can use the following instructions:

1. Install the (if prior to 1.5.0 being released)

    **If tedge 1.5.0 has not been released**

    ```sh
    wget -O - thin-edge.io/install.sh | sh -s -- --channel main
    ```

    **If tedge 1.5.0 has been released**

    ```sh
    wget -O - thin-edge.io/install.sh | sh -s
    ```

    **Note**

    If you are running the `wget` command on the device that needs to use the proxy to install thin-edge.io, then you'll probably need to set the linux environment variables to be able to download it properly (but we haven't fully verified this yet).

    ```sh
    SERVER="http://<proxy_host>:<proxyport>"
    export http_proxy=$SERVER
    export https_proxy=$SERVER
    export HTTP_PROXY=$SERVER
    export HTTPS_PROXY=$SERVER

    # Then run the install command again.
    wget -O - thin-edge.io/install.sh | sh -s
    ```

1. Disconnect the mapper and stop the tedge-agent service

    ```sh
    sudo tedge disconnect c8y
    sudo systemctl stop tedge-agent
    ```

1. Configure the proxy and switch to using the built-in tedge-mapper bridge

    ```sh
    # example http proxy
    sudo tedge config set proxy.address http://192.168.8.10:8080

    # Or if you are using a https proxy
    sudo tedge config set proxy.address https://192.168.8.10:8080
    ```

    If the proxy also requires authentication, then you can set the username/password using:

    ```sh
    sudo tedge config set proxy.username "myuser"
    sudo tedge config set proxy.password "mypasswrd"
    ```

1. Connect the tedge mapper to Cumulocity

    ```sh
    sudo tedge connect c8y
    ```
